### ISSUE TITLE
Change Request: `/* exported */` should not allow values

### ISSUE THREAD

**Issue by mdjermanovic** [eslint/eslint#17622]
### ESLint version

v8.50.0

### What problem do you want to solve?

Linter currently uses [`parseStringConfig`](https://github.com/eslint/eslint/blob/ee5be81fa3c4fe801c2f653854f098ed6a84dcef/lib/linter/config-comment-parser.js#L45) for parsing `/* exported */` comments.

`parseStringConfig` allows `name` and `name: value` entries.

When processing `/* exported */` comments, Linter uses names as the names of exported variables and just ignores values.

https://github.com/eslint/eslint/blob/ee5be81fa3c4fe801c2f653854f098ed6a84dcef/lib/linter/linter.js#L190-L198

This could be confusing, as in the following example:

```js
/* eslint no-unused-vars: "error" */

/* exported a: true, b: false */

var a, b; // both ok for no-unused-vars as they're considered exported
```

[Playground demo](https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXVudXNlZC12YXJzOiBcImVycm9yXCIgKi9cblxuLyogZXhwb3J0ZWQgYTogdHJ1ZSwgYjogZmFsc2UgKi9cblxudmFyIGEsIGI7IC8vIGJvdGggb2sgZm9yIG5vLXVudXNlZC12YXJzIGFzIHRoZXkncmUgY29uc2lkZXJlZCBleHBvcnRlZCIsIm9wdGlvbnMiOnsiZW52Ijp7ImVzNiI6dHJ1ZX0sInBhcnNlck9wdGlvbnMiOnsiZWNtYUZlYXR1cmVzIjp7fSwiZWNtYVZlcnNpb24iOiJsYXRlc3QiLCJzb3VyY2VUeXBlIjoic2NyaXB0In0sInJ1bGVzIjp7fX19)

### What do you think is the correct solution?

Linter should use [`parseListConfig`](https://github.com/eslint/eslint/blob/ee5be81fa3c4fe801c2f653854f098ed6a84dcef/lib/linter/config-comment-parser.js#L136) for parsing` /* exported */` comments and thus allow only `name` entries. 

### Participation

- [X] I am willing to submit a pull request for this change.

**Comment by amondev** [comment]
May I handle this issue? please assign to me~ I'll create PR soon!

**Comment by nzakas** [comment]
@amondev yes, thank you!

**Comment by mdjermanovic** [comment]
> So, we probably want to check for that case and avoid creating that variable, plus emit a warning?

Makes sense to check for that case, since it's currently valid so warnings might help with fixing the code for v9.

In the future, we could maybe add a more general check that would itself include this specific one, like whether the parsed value is a valid variable name or if the variable was actually found (`/* exported */` doesn't create variables; it just sets properties on existing ones and ignores non-existing ones).

https://github.com/eslint/eslint/blob/994596b07f5ff20a615a4be1ea03e5fd59cdb84b/lib/source-code/source-code.js#L308-L319

**Comment by nzakas** [comment]
So sounds like we want to check that a given exported variable is valid, and if not, emit a warning.

@amondev we probably want to use a regular expression, something like `/[^a-z0-9$_]+$/` on each variable name, and if it fails, emit a warning (similar to [this](https://github.com/eslint/eslint/blob/3aec1c55ba2c6d2833e1c0afe0a58f0cc6bbc0a4/lib/linter/linter.js#L423-L427).)

**Comment by amondev** [comment]
@nzakas I agree. Thank you for your detailed comments. I'll update PR soon.

**Comment by mdjermanovic** [comment]
If we assume that comments starting with the word "exported" are intended for ESLint, then it might be more useful to check if the variables they target actually exist in the global scope known to ESLint because otherwise the comment has no effect.

However, whichever check we choose, there's a concern that we'll be reporting comments that were not intended for ESLint, like we had in https://github.com/eslint/eslint/issues/14960. For example:

```js
/* exported because we need it in tests */
export { foo };
```

From this comment, `parseListConfig` will extract `"because we need it in tests"` value, which isn't a valid variable name so this comment would be reported but it wasn't intended for ESLint.

**Comment by mdjermanovic** [comment]
I just realized that `parseStringConfig` also allows space-separated values:

```js
/* eslint no-unused-vars: "error" */

/* exported a b */

var a, b; // both valid

```

But `parseListConfig` doesn't (values must be comma-separated), so `/* exported a b */` will also become invalid/no-effect if we switch to `parseListConfig`.

**Comment by nzakas** [comment]
> But parseListConfig doesn't (values must be comma-separated), so `/* exported a b */` will also become invalid/no-effect if we switch to parseListConfig.

If we don't have any tests validating `/* exported a b */` then I think it's fine to drop support for that. Arguably, forcing commas would reduce the likelihood that we'd collide with a comment that wasn't intended to mark a variable as exported.

Let's also keep in mind that usage of `/*exported*/` is likely incredibly low at this point. Both CommonJS and ESM use syntax to mark exports, so we're really talking about a minority of people using some other module system and needing to mark `var` declared variables so we don't complain that they are unused.

**Comment by amondev** [comment]
> But parseListConfig doesn't (values must be comma-separated), so /* exported a b */ will also become invalid/no-effect if we switch to parseListConfig.
 
Comment of `parseStringConfig` indicate that it can have names with whitespace. 🤔 

https://github.com/eslint/eslint/blob/3aec1c55ba2c6d2833e1c0afe0a58f0cc6bbc0a4/lib/linter/config-comment-parser.js#L39-L40

When we keep using `parseStringConfig`, `exportedVariables` contains both key and value.
What if we check both of them?

```
const variable = globalScope.set.get(key);
if (variable && (value === null || value === 'true' )) {
  variable.eslintUsed = true;
  variable.eslintExported = true;
}
```

**Comment by nzakas** [comment]
> When we keep using parseStringConfig, exportedVariables contains both key and value.
What if we check both of them?

Sorry, I'm not sure what it is you're asking here. Can you explain further?

**Comment by nzakas** [comment]
Ping @amondev

**Comment by mdjermanovic** [comment]
> If we don't have any tests validating `/* exported a b */` then I think it's fine to drop support for that.

We don't have any tests with space-separated variable names in `/* exported */` comments. In all tests that have multiple variable names in `/* exported */` comments, those variable names are comma-separated.

In the docs, all examples of `/* exported */` comments are with a single variable name.

In the example in the original issue https://github.com/eslint/eslint/issues/1200, variable names are comma-separated. In PR https://github.com/eslint/eslint/pull/2942 that introduced this feature, In all tests that have multiple variable names in `/* exported */` comments those variable names are comma-separated.

**Comment by mdjermanovic** [comment]
Since jshint was mentioned in the original issue (https://github.com/eslint/eslint/issues/1200#issuecomment-90939199). I also checked how it works and it seems to expect only comma-separated variable names.

```js
/*jshint unused:true*/
/* exported a, b */

var a = 5, b = 6; // ok
```

```js
/*jshint unused:true*/
/* exported a b */

var a = 5, b = 6; // 'a' is defined but never used. 'b' is defined but never used.
```

```js
/*jshint unused:true*/
/* exported a:true */

var a = 5; // 'a' is defined but never used.
```

**Comment by mdjermanovic** [comment]
> Let's also keep in mind that usage of `/*exported*/` is likely incredibly low at this point. Both CommonJS and ESM use syntax to mark exports, so we're really talking about a minority of people using some other module system and needing to mark `var` declared variables so we don't complain that they are unused.

Given this, and https://github.com/eslint/eslint/issues/17622#issuecomment-1781020710, I think it's fine that we just drop support for `/*exported a:true*/` and `/*exported a b*/` without any warnings. If we ever switch to a format that is clearly eslint-specific (e.g., `/* eslint-exported a */`) then we could emit warnings for comments that have no effect.

**Comment by nzakas** [comment]
> > Let's also keep in mind that usage of `/*exported*/` is likely incredibly low at this point. Both CommonJS and ESM use syntax to mark exports, so we're really talking about a minority of people using some other module system and needing to mark `var` declared variables so we don't complain that they are unused.
> 
> Given this, and [#17622 (comment)](https://github.com/eslint/eslint/issues/17622#issuecomment-1781020710), I think it's fine that we just drop support for `/*exported a:true*/` and `/*exported a b*/` without any warnings. If we ever switch to a format that is clearly eslint-specific (e.g., `/* eslint-exported a */`) then we could emit warnings for comments that have no effect.

Sounds good to me.

So to summarize, it seems like we've agreed that we should only support `/*exported a, b*/`, and we should not emit a warning for any other syntax.

**Comment by injae-kim** [comment]
> [OpenSource Study](https://chip-bream-9d5.notion.site/c96f1ae9aef840789d425b7a985b4eb5?v=0b33cfd3e31145eeb8d419be01cd4704) at GDG songdo, south korea (Help developers to make first PR on opensource they want)

JFYI) Hi @nzakas , @mdjermanovic , I'm [OpenSource Study](https://chip-bream-9d5.notion.site/c96f1ae9aef840789d425b7a985b4eb5?v=0b33cfd3e31145eeb8d419be01cd4704) leader(also contributor) and helping @amondev to fix this issue!

Cause @amondev has been busy these days for some personal reason, I'll ping to @amondev and update PR based on your discussion within next weekend~!

Really thanks a lot for your detailed discussions. 🙇

**Comment by amondev** [comment]
@nzakas @mdjermanovic I apologize for the delay in my response, which was due to personal reasons.
I'll update the PR along with @injae-kim soon!
